botocore/useragent.py

```diff
@@ -94,11 +141,40 @@ def to_string(self):
             self.name, allow_hash=False
         )
         if self.value is None or self.value == '':
-            return f'{clean_prefix}/{clean_name}'
-        clean_value = sanitize_user_agent_string_component(
-            self.value, allow_hash=True
-        )
-        return f'{clean_prefix}/{clean_name}#{clean_value}'
+            clean_string = f'{clean_prefix}/{clean_name}'
+        else:
+            clean_value = sanitize_user_agent_string_component(
+                self.value, allow_hash=True
+            )
+            clean_string = f'{clean_prefix}/{clean_name}#{clean_value}'
+        if self.size_config is not None:
+            clean_string = self._truncate_string(
+                clean_string,
+                self.size_config.max_size_in_bytes,
+                self.size_config.delimiter,
+            )
+        return clean_string
+
+    def _truncate_string(self, string, max_size, delimiter):
+        """
+        Pop ``delimiter``-separated values until encoded string is less than or
+        equal to ``max_size``.
+        """
+        orig = string
+        encoded_string = string.encode('utf-8')
+        while len(encoded_string) > max_size:
+            parts = string.split(delimiter)
+            parts.pop()
+            string = delimiter.join(parts)
+            encoded_string = string.encode('utf-8')
+
+        if string == '':
+            raise ValueError(
```

Comment:
Is raising an error (and failing the request) actually the behavior we want in this case? That seems wrong

Reply:
I'm thinking the behavior we want to experience in this _hopefully_ rare case is the following:
- Don't add the empty component to the user agent string
- Add a debug log statement explaining the issue/reason for omitting the UA component.

We should double check if this violates any requirements of the UA spec(s).